### PR TITLE
feat: use callback to parse TriggerSmartContract.data

### PR DESCRIPTION
**Pull request by andelf**
- loose length restriction, ok for 4 args
- allow unknown trc20 contract

### CODE REVIEW

**Comment by andelf** [comment]
test data(TriggerSmartContract with 3 args):

```
testdata: 0A02C35922087262BD3AF088C2554080B9D5A7922E5AD701081F12D2010A31747970652E676F6F676C65617069732E636F6D2F70726F746F636F6C2E54726967676572536D617274436F6E7472616374129C010A1541340967E825557559DC46BBF0EABE5CCF99FD134E1215418785CDFA387215DF8EDCFF4E0D7EA76E2110627B188080E983B1DE162264D3140A65000000000000000000000000856204182972E76AFFE436EFFA619511380ADE6A00000000000000000000000075B339AFA82A6FAD0E4216B85D4B07F8F364AA48000000000000000000000000000000000000000000000000016345785D8A000070B2EFD1A7922E90018094EBDC03
```

File: src/parse.c [hunk 1]
```diff
@@ -485,56 +485,80 @@ static bool account_update_contract(txContent_t *content,
...
+  content->decimals[0] = trc20->decimals;
+  content->tokenNamesLength[0] = strlen((const char *)trc20->ticker) + 1;
+  memmove(content->tokenNames[0], trc20->ticker, content->tokenNamesLength[0]);
+
+  return true;
+}
+
+static bool trigger_smart_contract(txContent_t *content, pb_istream_t *stream) {
+  msg.trigger_smart_contract.data.funcs.decode = pb_decode_trigger_smart_contract_data;
+  msg.trigger_smart_contract.data.arg = &content;
```
Comment (shydesky) [inline src/parse.c]: &content -> content

File: src/parse.c [hunk 2]
```diff
@@ -485,56 +485,80 @@ static bool account_update_contract(txContent_t *content,
...
-  if (TRC20 == NULL) {
+
+  // amount
+  if (!pb_read(stream, buf, 32)) {
+    return false;
+  }
+
+  memmove(content->TRC20Amount, buf, 32);
+  tokenDefinition_t *trc20 = getKnownToken(content);
+  if (trc20 == NULL) {
```
Comment (fbsobreira) [inline src/parse.c]: Maybe would be good to add an option, if custom contract is set true we can authorize unknown tokens, but not showing amount...

Comment (andelf) [reply src/parse.c]: Mark this as a TODO. 🤪

**Pull request merged**